test: DatabaseSelector | Splitting the file and creating tests

yardz opened the pull request:
### SUMMARY
The DatabaseSelector component was very complex and had several responsibilities, in addition to developing several rerenders that were not necessary.

#### First Step
The first step was to separate the functions that were inside the component into hooks. Leaving this out we have some direct improvements:
- Improved readability (principle of single responsibility and cleaner code)
- Allow use of useRef to prevent unnecessary rerenders
- Test possibility

For the functions, the same structure is followed. There is a factory, which takes as an argument functions that will be used within the original function (usually the "setSomething") and returns a function with the same input and behavior as the original function. This factory is used to initialize a `useRef` and has a `useEffect` to update the refs when any prop is changed.
This way, the component only renders again when there is a need to update the UI, but the functions are always with the current props and states.

#### Second step
There were still 4 components inside `DatabaseSelector`, using names like `renderDatabaseSelect`.
Each of these has been separated into a single component. Facilitating testing, making the code cleaner, and facilitating future maintenance.
Even by separating the components of the behavior it was maintained and the interface was kept as similar as possible.


### TEST PLAN
No behavior change in the application. All tests must pass correctly.

yardz committed:
Moving DatabaseSelector to a folder

yardz committed:
Switching fetchSchemas function to a hook

yardz committed:
Switching onSelectChange to a hook

yardz committed:
Switching dbMutator to a hook

yardz committed:
Switching changeDataBase to a hook

yardz committed:
Switching changeSchema to a hook

yardz committed:
Moving DatabaseOption to a component

yardz committed:
Moving SelectRow to a component

yardz committed:
Moving SchemaSelect to a component

yardz committed:
fix: License Comments

yardz committed:
split getQueryParams

yardz committed:
Component DatabaseSelect

yardz committed:
Tests for DatabaseSelector

yardz committed:
Merge remote-tracking branch 'apache/master' into test-DatabaseSelector

yardz committed:
Applying commit changes #13409 (style)

yardz committed:
fix: SchemaSelect test

michael-s-molina commented:
Hey @yardz this is a big one! Since you broke the component into many parts I'll try to break the review also 😆. 

I'll focus more on general topics instead of code lines:
- You did a great job with typescript support. All files are in typescript and the types are well defined.
- The tests also look great and this will definitely increase test coverage.
- I do have a concern about introducing regressions because of the changes. We definitely need to execute acceptance tests here.

The next part is subjective, there is no right or wrong, so feel free to apply any recommendation:
- I think we can slightly increase the limit for breaking a component. Some components are very simple, non-reusable, and appear to be an internal part of the parent component. We don't want the original size of the component but also don't want to end up with labels or simple divs as separate components. A lot of fragmentation also doesn't help with readability. The key here is **balance**.
- Hooks are essentially **reusable** functions. You generally create them when you want to share stateful logic between components. Some of the hooks that you created are not reusable but internal functions of this component. I think they belong to the subcomponents that you created.

As I said these are subjective, so feel free to apply any 😉

michael-s-molina commented:
@yardz As I said those are subjective recommendations. For me, what I learned from experience is that the extremes are never the right choice. So creating a gigantic component or breaking it at the smallest level possible is not the optimal solution. As I said earlier, the key is balance. In my opinion, readability is affected in both cases.

ktmud commented:
I don't understand how a 290 LOC file can be expanded to 2000+ lines. How is this actually making anything simple?

rusackas commented:
Agreed with above statements. I understand and agree with the notion that complex components should be broken down _as needed_ to support higher testability.

That said, this swings the pendulum too far in the other direction. I don't think 38 files and 2K lines for this component is the right approach either.

I'll try to go through this piece by piece and give thoughts as to which files seem necessary (or not) to break up.

I think in general, we can cut a component in half as one PR, and then subdivide as needed in subsequent PRs. Breaking a mountain into boulders is one thing, but turning the mountain into sand might be going too far :)

Clearly a lot of diligent work went into this PR, so I want to treat the situation respectfully and carefully. Long story short, I'd rather have a file with a thousand lines open, rather than open 20+ IDE tabs and figure out how it all connects.

suddjian commented on superset-frontend/src/components/DatabaseSelector/changeDataBase/useChangeDataBase.ts:
```diff
@@ -0,0 +1,64 @@
+      onSelectChange,
+    }),
+  );
+  useEffect(() => {
```
Could this be implemented with a `useCallback` instead, moving the contents of `factoryChangeDataBase` into here? This is a lot of ref usage, which gets confusing and doesn't seem very idiomatic to hooks.

yardz replied:
`useCallback` does not tackle the problem the way I wanted it. It will generate a new function every time a parameter changes and will cause a new render.
As there are several functions that depend on each other, several rendes will be caused by this. When the `current` value is updated, a render is not trigged (which is why I don't return the `current` either).

The idea there is to behave like a class property. I want to change the value of these functions without triggering a new render. And of course, from the moment that this value is changed, everyone who uses this function (and has received these props in the past) will access the new value without a new rendering.

The combination of `useEffect` + `useRef` allows me to do this. Using `useCallback` or `useMemo` does not do this.

suddjian commented on superset-frontend/src/components/DatabaseSelector/changeDataBase/useChangeDataBase.ts:
```diff
@@ -0,0 +1,64 @@
+    setSchemaOptions,
+  ]);
+
+  return changeDataBase;
```
Also, the caller of this hook has to do `changeDataBase.current()`. If we continue using a ref for this, it should return `changeDataBase.current` instead.

suddjian commented on superset-frontend/src/components/DatabaseSelector/changeSchema/useChangeSchema.ts:
```diff
@@ -0,0 +1,54 @@
+      onSelectChange,
+    }),
+  );
+  useEffect(() => {
```
Same comment on `useRef` + `useEffect` vs `useCallback`

yardz replied:
Very different results. I explained in the first comment.

ktmud commented:
> > I don't understand how a 290 LOC file can be expanded into 2000+ lines. Even excluding tests and LICENSE blurbs, it is at least 3x more code to understand. How is this actually making anything simpler?
> > Code structure is important. Code amount also matters!
> 
> When you put 2000 lines I was scared, since I didn't actually write much code.
> 
> Actually it didn't even come close to that, I took all the code and put it in 1 file, removed the license and the signatures of the functions. There are 240 lines of code. I did the same with the original code and there are 190 lines. I believe that this overhead is due to licenses and functions declarations.
> 
> (only the tests are 1000 lines)

Did you count all the hooks you created though?  I did the same and got 21 non-test files and 700 LOC: https://gist.github.com/ktmud/f5c16eb18bbceda2fd3f998d14f18948

It was almost 700 lines vs 270 lines. Better than my original estimate, but still... it's also 21 files vs 1 file... I can't imagine opening 21 files just trying to understand this one component.

yardz commented:
> Did you count all the hooks you created though? I did the same and got 21 non-test files and 700 LOC: https://gist.github.com/ktmud/f5c16eb18bbceda2fd3f998d14f18948

Yes, but I just compared the "code body". I removed the imports and the declaration of methods like this:
```
Removed:
--> import React, { useEffect, useRef } from 'react';
--> 
--> import { OnSelectChange } from '../onSelectChange';
--> import { factoryChangeSchema } from './factoryChangeSchema';
--> 
--> export const useChangeSchema = ({
-->   setCurrentSchema,
-->   onSchemaChange,
-->   getTableList,
-->   onSelectChange,
--> }: {
-->   setCurrentSchema: (schema: string | undefined) => void;
-->   onSchemaChange?: (arg0?: any) => {};
-->   getTableList?: (dbId: number, schema: string, force: boolean) => {};
-->   onSelectChange: React.MutableRefObject<OnSelectChange>;
--> }) => {
-->   const changeSchema = useRef(
-->     factoryChangeSchema({
-->       setCurrentSchema,
-->       onSchemaChange,
-->       getTableList,
-->       onSelectChange,
-->     }),
-->   );
  useEffect(() => {
    changeSchema.current = factoryChangeSchema({
      setCurrentSchema,
      onSchemaChange,
      getTableList,
      onSelectChange,
    });
  }, [getTableList, onSchemaChange, setCurrentSchema, onSelectChange]);

  return changeSchema;
};
```
How I basically moved existing code to an external function.

> It was almost 700 lines vs 270 lines. Better than my original estimate, but still... it's also 21 files vs 1 file... I can't imagine opening 21 files just trying to understand this one component.

21 are with the test files. But yes, they are much more files.

ktmud commented:
21 is without the test files.
